Fixes #38024 - Add ordering by virtual columns

ofedoren opened the pull request:
This is an alternative for https://github.com/theforeman/foreman/pull/10368, tries to make ordering by columns, which are not part of actual table, possible.

This PR enables it for roles model, making possible to click on `locked` column in roles index page and see the ordering, as well as doing so via API.

Added an entry for developer_docs saying how to use it within plugins.

Thanks, @adamruzicka, for the idea!

@MariaAga, what do you think?

P.S. The only issue I see is that we'd need to update https://github.com/theforeman/foreman-tasks/blob/master/app/controllers/foreman_tasks/api/tasks_controller.rb#L195 to be compatible. Could be done as part of removal deprecated stuff.

MariaAga commented:
It seems to work as expected in ui and api, the code makes sense but I think we need a Ruby person to verify it, @adamruzicka ?

adamruzicka commented on app/controllers/application_controller.rb:
```diff
@@ -397,6 +397,16 @@ def parameter_filter_context
+    if virt_column.blank? || model_of_controller.columns_hash[virt_column] || !base_scope.respond_to?(select_method)
+      base_scope.search_for(params[:search], :order => params[:order]).paginate(:page => params[:page], :per_page => params[:per_page])
+    else
+      base_scope.send(select_method).search_for(params[:search]).order(params[:order]).paginate(:page => params[:page], :per_page => params[:per_page])
```
UI does not accept per_page being set to all?

adamruzicka replied:
Also, thinking out loud, from the name I wouldn't really expect it to deal with pagination and so on, but that might be just me

ofedoren replied:
> UI does not accept per_page being set to all?

Nope, otherwise we'd have a button suggesting that.

> Also, thinking out loud, from the name I wouldn't really expect it to deal with pagination and so on, but that might be just me

Naming is hard, any suggestion is more than welcome :cookie:

MariaAga replied:
Does this break per_page=all for api?

ofedoren replied:
> Does this break per_page=all for api?

`wrap_for_virt_column_select` method is only for UI controllers, thus no.

API controllers most (if not all) of the time use `resource_scope_for_index` method, which is overridden in this PR:
https://github.com/theforeman/foreman/pull/10381/files#diff-f97adc72a387ed049f1b789f93f9d5c592f481e9e37c6fb75eabfcc2f36d467eR172

This override should not affect `per_page=all`, since it's either uses `super` to use old behavior or explicitly deals with it in https://github.com/theforeman/foreman/pull/10381/files#diff-f97adc72a387ed049f1b789f93f9d5c592f481e9e37c6fb75eabfcc2f36d467eR181.

adamruzicka reviewed (commented):
Looks reasonable, left two comments. Could we also get a related patch in one of the plugins where its needed?

ofedoren committed:
Fixes #38024 - Add ordering by virtual columns

Refs #38024 - Check if select_virt_col is allowed

Refs #38024 - Refactor the same logic

Refs #38024 - Reorder in case default scope with order

ofedoren commented:
I've updated the logic a bit: since most of the UI controllers use `resource_base_search_and_page`/`resource_base_with_search`, I updated this method, so the new `wrap_for_virt_column_select` method is not necessary, but rather an optional helper. Although, looking at the changes in https://github.com/theforeman/foreman_openscap/pull/587, I don't think this wrapper will be used...

WDYT? Should we expose the wrapper or rather force usage of already defined loaders if ordering by virt columns desired?

ofedoren commented:
@adamruzicka, @MariaAga, reviving this one: do we still want that?

Note: it also fixes broken ordering by `locked` on `roles` page.

ofedoren commented on app/controllers/application_controller.rb:
```diff
@@ -232,7 +232,13 @@ def remote_user_provided?
+    select_method = "select_#{virt_column}"
+    if virt_column.blank? || model_of_controller.columns_hash[virt_column] || !resource_base.respond_to?(select_method)
+      resource_base.search_for(params[:search], :order => params[:order])
+    else
```
Applied as https://github.com/theforeman/foreman/pull/10381/commits/3776040a096b3900c5b681be3999cd6f43d89eb1.

adamruzicka commented on app/controllers/application_controller.rb:
```diff
@@ -232,7 +232,13 @@ def remote_user_provided?
+    if virt_column.blank? || model_of_controller.columns_hash[virt_column] || !resource_base.virtual_column_scopes.include?(select_method.to_sym)
+      resource_base.search_for(params[:search], :order => params[:order])
+    else
+      resource_base.public_send(select_method).search_for(params[:search]).order(params[:order])
```
Is there a difference between passing the order to `#search_for` and passing it to chained `#order`?

ofedoren replied:
It's to workaround https://github.com/wvanbergen/scoped_search/blob/master/lib/scoped_search/definition.rb#L332, which seems to ignore virtual column.

adamruzicka replied:
Oh, right, forgot about that

adamruzicka commented on app/controllers/api/v2/base_controller.rb:
```diff
@@ -168,6 +168,20 @@ def self.hide_taxonomy_options
+        select_method = "select_#{virt_column}"
+        scope = resource_scope(...)
+
+        if virt_column.blank? || resource_class.columns_hash[virt_column] || !scope.virtual_column_scopes.include?(select_method.to_sym)
```
We have this pattern with small differences in three different places, could we make it dryer?

ofedoren replied:
Done. It indeed looks better now.

adamruzicka commented:
/packit build

ofedoren commented on app/controllers/concerns/application_shared.rb:
```diff
@@ -150,6 +150,6 @@ def virtual_column_scope(base_scope)
     model = respond_to?(:model_of_controller) ? model_of_controller : resource_class
     return if virt_column.blank? || model.columns_hash[virt_column] || !base_scope.virtual_column_scopes.include?(select_method.to_sym)
-    base_scope.public_send(select_method).search_for(params[:search]).order(params[:order])
+    base_scope.public_send(select_method).search_for(params[:search]).reorder(params[:order])
```
This already considers `unless params[:order].blank?` since if it's empty the method returns earlier and the logic returns to the previous (current) state ignoring virtual column.

adamruzicka commented:
Could you rebase please?

adamruzicka reviewed (approved):
Tested together with the foreman_openscap PR, works well

adamruzicka commented:
Could you please squash the commits?

ofedoren commented:
Thanks, @adamruzicka, rebased / squashed. Although I'd like to see green tests just in case, but I don't think there will be related failures anyway :shrug:

ofedoren commented:
:green_apple:

adamruzicka commented:
Thank you @ofedoren !
